src/liboslexec/backendllvm.cpp

```diff
@@ -267,13 +268,25 @@ BackendLLVM::llvm_alloca(const TypeSpec& type, bool derivs,
 }
 
 
+bool
+BackendLLVM::is_stack_parameter(const Symbol& sym)
+{
+    if (!shadingsys().m_opt_groupdata)
+        return false;
+
+    // Some output parameters that are never needed before or
+    // after layer execution can be relocated from GroupData
+    // onto the stack.
+    return sym.symtype() == SymTypeOutputParam && !sym.renderer_output()
+           && !sym.typespec().is_closure_based() && !sym.connected();
```

Comment:
Question for my understanding: You don't allow any type of connection? I thought downstream was allowed.

Reply:
Thanks for taking a look! Symbol::connected() is upstream only, there's a separate method connected_down() for downstream connections.

Reply:
I see, thanks!

Reply:
Does is_stack_parameter take into account "lockgeom()==0" or "interpolated" or reparameter into account?  Does it need to (maybe not)?

Reply:
That's a good question about input parameters. I took a look at how OSL is currently optimizing 4 cases of unconnected parameters:

```
float u_func() { return u; }

shader opt_candidates(
    float w = 0.,
    float x = u,
    float y = u_func(),
    float z = 0. [[ int lockgeom = 0 ]]
)
{...}
```
Currently two of these cases are optimized out of the GroupData:
* `w`: constant-folded 
* `x`: I'm not sure if constant-folding is still the correct term here, but x is optimized out

The other two aren't:
* `y`: I hoped y would optimize like x, but it doesn't
* `z`: triggers a user-data lookup, so not constant-folded

I updated my patch to catch `z` (I believe unconnected + interpolated is safe for stack placement). It passed the unit tests but led to some image differences internally. I'll stick to output parameters for now, hopefully I can figure out that bug and follow up with input parameters.

Reply:
You might also want to examine behavior with `-O0` or `setenv TESTSHADE_OPT 0` that might cause that x to be left around as well as an input.